### ISSUE TITLE
Service stop responding after dropped SSE connections (Only on Linux)

### ISSUE THREAD

**Issue by REDDOXX** [ServiceStack/Issues#391]
We have this issue since more than a year, but it took a really long time to found the real reason.
After ~1000 dropped SSE connections the service will not longer respond to any connection at all.

It took some time but now i was able to provide you a sample to reproduce this issue.
The problem only occurs on Linux, and you can reproduce it with any version of Mono incl. 4.2.1

I made a sample project on Github with detailed description. So you should be able to to reproduce the issue very fast. Hopefully you find a way to fix this issue, because it provides an easy way to DDoS a ServiceStack service with SSE.

You can find the Sample-Project on Github:
https://github.com/REDDOXX/ServiceStack-SSE-Issue

Alex

**Comment by mythz** [comment]
This is an issue with Mono's HTTP Stack in general which is known to have memory and performance issues under heavy load. Mono's Load issues aren't something that can be resolved within ServiceStack and why we [don't recommend Mono for heavy workloads](https://plus.google.com/u/1/+RyanBritton/posts/E98iKBTLTB7).

Unfortunately the only future we see for running heavy .NET HTTP production workloads on Linux is on the cross-platform CoreCLR which Microsoft will support and maintain in future. We'll look into the effort of supporting DNX after it's released.

**Comment by REDDOXX** [comment]
No. This issue has nothing to do with heavy load.
You can reproduce the issue even if you make one connection per second, or one per minute.
Every(!) dropped connection will leave one leaked connection (socket), and after ~1000 dropped connections the service is dead.

The issue only occurs for dropped requests to SSE e.g. http://localhost:1337/event-stream
I never see any problems with "normal" service requests.

Alex

**Comment by mythz** [comment]
I'm seeing leaking sockets reported in a few places for Mono, e.g: https://github.com/SignalR/SignalR/issues/3161

That thread mentions a potential workaround by setting `HttpListener.IgnoreWriteExceptions`. Can you change your AppHost to include:

```csharp
public override ServiceStackHost Start(string urlBase)
{
    base.Start(urlBase);
    Listener.IgnoreWriteExceptions = true;
    return this;
}
```

Can you let me know if that helps?

I've managed to exceed 2000+ connections killed with `./kill-connections.sh` with this enabled. Note I've modified it to `COUNT_PER_SEC=1`.

Incidentally `./monitor.sh` is always reporting:

    0 leaked connections found.

But when I did `lsof -p $PID` on its own I was able to see a growing number of `(CLOSED)` sockets initially, but stopped growing after `IgnoreWriteExceptions=true`.

I'm running this on OSX El Capitan 10.11.2 (15C50) / Mono 4.2.1

**Comment by REDDOXX** [comment]
This will help but is not optimal.
With this option the socket will not longer leak, but the socket will hold until the ServerEventsFeature.IdleTimeout was be reached (Default 30 seconds).
After the timeout is reached, the socket will be closed.

I spent the hole day in debugging this, and hopefully found a better solution:

Change the Publish method in the class EventSubscription:

```C#
public void Publish(string selector, string message)
{
	try
	{
		var msg = message ?? "";
		var frame = "id: " + Interlocked.Increment(ref msgId) + "\n"
				  + "data: " + selector + " " + msg + "\n\n";

		lock (response)
		{
			WriteEvent(response, frame);   // <-- Here the exception occurs, if the connection was dropped

			if (OnPublish != null)
				OnPublish(response, frame);
		}
	}
	catch (Exception ex)
	{
		Log.Error("Error publishing notification to: " + selector, ex);
		
		// If we explicit close the OutputStream after the error, then the socket will not leak ( response.Close() will not work ! )
		try
		{
			// This will throw an exception, but on Mono (Linux / OSX) the socket will leak if we not close the OutputStream
			response.OutputStream.Close();
		}
		catch { }
		
		Unsubscribe();
	}
}
```

With this modification the dropped connection will be closed after the next Publish. In many cases this should be much faster than waiting for the IdleTimeout.
What do you think ?

Alex

**Comment by mythz** [comment]
ok cool, I've added the above Exception handling in this commit: https://github.com/ServiceStack/ServiceStack/commit/e414741719f1d873d09681e1f891e855d67cfffc

This change is available from v4.0.53 that's now [available on MyGet](https://github.com/ServiceStack/ServiceStack/wiki/MyGet).

**Comment by REDDOXX** [comment]
Perfect. I've updated the sample to v4.0.53 and the sockets will not longer leak.
